src/resolution/graph.rs

```diff
@@ -1264,44 +1286,84 @@ impl<'a, TNpmRegistryApi: NpmRegistryApi>
-    for ancestor_node in ancestor_path.ancestors() {
-      match ancestor_node {
-        GraphPathNodeOrRoot::Node(ancestor_graph_path_node) => {
-          path.push(ancestor_graph_path_node);
-          let maybe_peer_dep = self.find_peer_dep_in_node(
-            ancestor_graph_path_node,
-            peer_dep,
-            peer_package_info,
-            None,
-            ancestor_path,
-          )?;
-          if let Some((parent, peer_dep_id)) = maybe_peer_dep {
-            // this will always have an ancestor because we're not at the root
-            self.set_new_peer_dep(&path, parent, specifier, peer_dep_id);
-            return Ok(Some(peer_dep_id));
+    let mut matching_peers_going_up = ancestor_path
+      .ancestors()
+      .enumerate()
+      .filter_map(|(i, ancestor_node)| {
+        match ancestor_node {
+          GraphPathNodeOrRoot::Node(ancestor_graph_path_node) => {
+            let maybe_peer_dep_result = self.find_peer_dep_in_node(
+              ancestor_graph_path_node,
+              peer_dep,
+              peer_package_info,
+              None,
+              ancestor_path,
+            );
+            match maybe_peer_dep_result {
+              Ok(maybe_peer_dep) => {
+                maybe_peer_dep.map(|(peer_parent, peer_dep_id)| {
+                  let path = get_path(i, ancestor_path);
+                  Ok((path, peer_parent, peer_dep_id))
+                })
+              }
+              Err(err) => Some(Err(err)),
+            }
           }
-        }
-        GraphPathNodeOrRoot::Root(root_pkg_nv) => {
-          // in this case, the parent is the root so the children are all the package requirements
-          if let Some(child_id) = self.find_matching_child_for_peer_dep(
-            peer_dep,
-            peer_package_info,
-            self.graph.root_packages.iter().map(|(nv, id)| (*id, nv)),
-            ancestor_path,
-          )? {
-            let peer_parent = GraphPathNodeOrRoot::Root(root_pkg_nv.clone());
-            self.set_new_peer_dep(&path, peer_parent, specifier, child_id);
-            return Ok(Some(child_id));
+          GraphPathNodeOrRoot::Root(root_pkg_nv) => {
+            // in this case, the parent is the root so the children are all the package requirements
+            let maybe_peer_dep_result = self.find_matching_child_for_peer_dep(
+              peer_dep,
+              peer_package_info,
+              self.graph.root_packages.iter().map(|(nv, id)| (*id, nv)),
+              ancestor_path,
+            );
+            match maybe_peer_dep_result {
+              Ok(maybe_peer_dep) => maybe_peer_dep.map(|peer_dep_id| {
+                let path = get_path(i, ancestor_path);
+                let peer_parent =
+                  GraphPathNodeOrRoot::Root(root_pkg_nv.clone());
+                Ok((path, peer_parent, peer_dep_id))
+              }),
+              Err(err) => Some(Err(err)),
+            }
           }
         }
+      });
```

Comment:
By the way, I tried really hard to extract this out to a separate function, but I could get the code to compile (and I'm pretty sure I was expressing it correctly with lifetimes).